File: worker/src/aggregate/workload.ts
export type StatusBucket = "backlog" | "inProgress" | "inReview" | "closed";

export type IssueLike = {
  key: string;
  statusName: string;
  statusCategory?: string | null;
  assigneeName: string | null;
};

export type AssigneeWorkload = {
  assignee: string;
  total: number;
  counts: Record<StatusBucket, number>;
  percents: Record<StatusBucket, number>;
};

const BUCKETS: StatusBucket[] = [
  "backlog",
  "inProgress",
  "inReview",
  "closed",
];

export function mapStatusToBucket(
  statusName: string,
  statusCategory?: string | null,
): StatusBucket {
  const name = statusName.trim().toLowerCase();
  const category = (statusCategory ?? "").trim().toLowerCase();

  if (/review/.test(name)) return "inReview";

  if (
    category === "done" ||
    /\b(done|closed|resolved|complete[d]?)\b/.test(name)
  ) {
    return "closed";
  }

  if (
    category === "indeterminate" ||
    /\b(in progress|progress|development|building|doing)\b/.test(name)
  ) {
    return "inProgress";
  }

  return "backlog";
}

/** Largest-remainder so the four percentages always sum to 100 (or 0 if empty). */
export function percentsFromCounts(
  counts: Record<StatusBucket, number>,
  total: number,
): Record<StatusBucket, number> {
  const empty = { backlog: 0, inProgress: 0, inReview: 0, closed: 0 };
  if (total <= 0) return empty;

  const raw = BUCKETS.map((b) => ({
    bucket: b,
    exact: (counts[b] / total) * 100,
  }));
  const floored = raw.map((r) => ({
    ...r,
    value: Math.floor(r.exact),
    frac: r.exact - Math.floor(r.exact),
  }));
  let remaining = 100 - floored.reduce((s, r) => s + r.value, 0);
  floored
    .slice()
    .sort((a, b) => b.frac - a.frac)
    .forEach((r) => {
      if (remaining <= 0) return;
      const target = floored.find((f) => f.bucket === r.bucket);
      if (target) {
        target.value += 1;
        remaining -= 1;
      }
    });

  return Object.fromEntries(
    floored.map((f) => [f.bucket, f.value]),
  ) as Record<StatusBucket, number>;
}

export function aggregateByAssignee(issues: IssueLike[]): AssigneeWorkload[] {
  const byAssignee = new Map<
    string,
    Record<StatusBucket, number> & { total: number }
  >();

  for (const issue of issues) {
    const assignee = issue.assigneeName?.trim() || "Unassigned";
    const bucket = mapStatusToBucket(issue.statusName, issue.statusCategory);
    let row = byAssignee.get(assignee);
    if (!row) {
      row = { backlog: 0, inProgress: 0, inReview: 0, closed: 0, total: 0 };
      byAssignee.set(assignee, row);
    }
    row[bucket] += 1;
    row.total += 1;
  }

  return [...byAssignee.entries()]
    .map(([assignee, row]) => {
      const counts = {
        backlog: row.backlog,
        inProgress: row.inProgress,
        inReview: row.inReview,
        closed: row.closed,
      };
      return {
        assignee,
        total: row.total,
        counts,
        percents: percentsFromCounts(counts, row.total),
      };
    })
    .sort((a, b) => b.total - a.total || a.assignee.localeCompare(b.assignee));
}

File: worker/app/aggregate/workload.py
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Literal

StatusBucket = Literal["backlog", "inProgress", "inReview", "closed"]

BUCKETS: tuple[StatusBucket, ...] = ("backlog", "inProgress", "inReview", "closed")


@dataclass(frozen=True)
class Issue:
    key: str
    status_name: str
    status_category: str | None
    assignee_name: str | None


@dataclass
class AssigneeWorkload:
    assignee: str
    total: int
    counts: dict[StatusBucket, int]
    percents: dict[StatusBucket, int]

    def to_dict(self) -> dict:
        return {
            "assignee": self.assignee,
            "total": self.total,
            "counts": self.counts,
            "percents": self.percents,
        }


def map_status_to_bucket(
    status_name: str,
    status_category: str | None = None,
) -> StatusBucket:
    name = status_name.strip().lower()
    category = (status_category or "").strip().lower()

    if "review" in name:
        return "inReview"

    if category == "done" or re.search(
        r"\b(done|closed|resolved|complete[d]?)\b", name
    ):
        return "closed"

    if category == "indeterminate" or re.search(
        r"\b(in progress|progress|development|building|doing)\b", name
    ):
        return "inProgress"

    return "backlog"


def percents_from_counts(
    counts: dict[StatusBucket, int],
    total: int,
) -> dict[StatusBucket, int]:
    """Largest-remainder so the four percentages always sum to 100 (or 0 if empty)."""
    empty: dict[StatusBucket, int] = {
        "backlog": 0,
        "inProgress": 0,
        "inReview": 0,
        "closed": 0,
    }
    if total <= 0:
        return empty

    floored: list[dict] = []
    for b in BUCKETS:
        exact = (counts[b] / total) * 100
        floored.append(
            {
                "bucket": b,
                "value": int(exact // 1),
                "frac": exact - int(exact // 1),
            }
        )

    remaining = 100 - sum(r["value"] for r in floored)
    for r in sorted(floored, key=lambda x: x["frac"], reverse=True):
        if remaining <= 0:
            break
        target = next(f for f in floored if f["bucket"] == r["bucket"])
        target["value"] += 1
        remaining -= 1

    return {f["bucket"]: f["value"] for f in floored}


def aggregate_by_assignee(issues: list[Issue]) -> list[AssigneeWorkload]:
    by_assignee: dict[str, dict[str, int]] = {}

    for issue in issues:
        assignee = (issue.assignee_name or "").strip() or "Unassigned"
        bucket = map_status_to_bucket(issue.status_name, issue.status_category)
        row = by_assignee.get(assignee)
        if row is None:
            row = {
                "backlog": 0,
                "inProgress": 0,
                "inReview": 0,
                "closed": 0,
                "total": 0,
            }
            by_assignee[assignee] = row
        row[bucket] += 1
        row["total"] += 1

    rows: list[AssigneeWorkload] = []
    for assignee, row in by_assignee.items():
        counts: dict[StatusBucket, int] = {
            "backlog": row["backlog"],
            "inProgress": row["inProgress"],
            "inReview": row["inReview"],
            "closed": row["closed"],
        }
        total = row["total"]
        rows.append(
            AssigneeWorkload(
                assignee=assignee,
                total=total,
                counts=counts,
                percents=percents_from_counts(counts, total),
            )
        )

    rows.sort(key=lambda r: (-r.total, r.assignee))
    return rows
